art/attacks/evasion/pe_malware_attack.py

```diff
@@ -0,0 +1,778 @@
+    3) Slack manipulation attacks (example paper link: https://arxiv.org/abs/1810.08280)
+    4) DOS Header Attacks (example paper link: https://arxiv.org/abs/1901.03583)
+"""
+
+import json
+import random
+import copy
+import sys
+
+from typing import Optional, Union, Tuple, List, Dict, TYPE_CHECKING
+
+import lief
+import numpy as np
+import tensorflow as tf
+
+from art.estimators.estimator import BaseEstimator, LossGradientsMixin
+from art.attacks.attack import EvasionAttack
+
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
+
+
+class MalwareGD(EvasionAttack):
+    """
+    Implementation of the following white-box attacks related to PE malware crafting:
+        1) Append based attacks (example paper link: https://arxiv.org/abs/1810.08280)
+        2) Section insertion attacks (example paper link: https://arxiv.org/abs/2008.07125)
+        3) Slack manipulation attacks (example paper link: https://arxiv.org/abs/1810.08280)
+        4) DOS Header Attacks (example paper link: https://arxiv.org/abs/1901.03583)
+    """
+
+    _estimator_requirements = (BaseEstimator, LossGradientsMixin)
+
+    def __init__(
+        self,
+        prediction_model: "CLASSIFIER_NEURALNETWORK_TYPE",
+        embedding_weights: np.ndarray,
+        param_dic: Dict[str, int],
+        num_of_iterations: int = 10,
+        l_0: Union[float, int] = 0.1,
+        l_r: float = 1.0,
+        use_sign: bool = False,
+        verbose: bool = False,
+    ) -> None:
+
+        """
+        :param prediction_model: A trained classifier that takes in the PE embeddings to make a prediction.
+        :param embedding_weights: weights for the embedding layer
+        :param param_dic: a dictionary specifying some MalConv parameters.
+                         'maxlen': the input size to the MalConv model
+                         'input_dim': the number of discrete values, normally 257.
+                         'embedding_size': size of the embedding layer. Default 8.
+        :param num_of_iterations: The number of iterations to apply.
+        :param l_0: l_0 bound for the attack. If less then 1 it is interpreted as a fraction of the file size.
+                    If larger than 1 it is interpreted as the total number of permissible features to change.
+        :param l_r: learning rate for the optimisation
+        :param use_sign: if we want to use the sign of the gradient, rather then the gradient itself.
+        :param verbose: if additional information will be printed to the terminal
+        """
+        super().__init__(estimator=prediction_model)
+        self.param_dic = param_dic
+        self.embedding_weights = embedding_weights.astype("float32")
+        self.l_0 = l_0
+        self.l_r = l_r
+        self.use_sign = use_sign
+        self.total_perturbation: np.ndarray = np.zeros(shape=(1, 1))
+        self.num_of_iterations = num_of_iterations
+        self.verbose = verbose
+
```

Comment:
Please add a method `_check_params` similar to the other evasion attacks.
```suggestion
        self.verbose = verbose
        self._check_params()

```